source/code/systems/resource_system/private/resource_tracker.cxx

```diff
@@ -44,431 +58,446 @@ namespace ice
+        ice::ResourceTrackerCreateInfo _info;
+
+        ice::TaskQueue _io_queue;
+        ice::UniquePtr<ice::NativeIO> _io_thread_data;
+        ice::UniquePtr<ice::TaskThread> _io_thread;
+
+        ice::Array<ice::ResourceHandle, ContainerLogic::Complex> _handles;
+        ice::HashMap<ice::ResourceHandle*> _resources;
+        ice::HashMap<ice::ResourceProvider*> _resource_providers;
+    };
+
+    ResourceTrackerImplementation::ResourceTrackerImplementation(
+        ice::Allocator& alloc,
+        ice::TaskScheduler& scheduler,
+        ice::ResourceTrackerCreateInfo const& info
+    ) noexcept
+        : ResourceTracker{ }
+        , _allocator{ alloc }
+        , _scheduler{ scheduler }
+        , _info{ info }
+        , _io_queue{ }
+        , _io_thread_data{ }
+        , _io_thread{ }
+        , _handles{ _allocator }
+        , _resources{ _allocator }
+        , _resource_providers{ _allocator }
+    {
+        ICE_ASSERT(
+            _info.predicted_resource_count > 0,
+            "Invalid value ({}) provided for 'predicted_resource_count'. Value needs to be a positive integer."
+        );
+
+        ice::array::reserve(_handles, _info.predicted_resource_count);
+        ice::hashmap::reserve(_resources, _info.predicted_resource_count);
+        ice::hashmap::reserve(_resource_providers, 10);
+
+        if (_info.io_dedicated_threads > 0)
         {
-            ice::hashmap::remove(_providers, ice::hash_from_ptr(provider));
-            return true;
+            _io_thread_data = ice::create_nativeio_thread_data(_allocator, _info.io_dedicated_threads);
+
+            ice::TaskThreadInfo const io_thread_info{
+                //.exclusive_queue = true, // ignored
+                //.sort_by_priority = false, // ignored
+                .custom_procedure = (ice::TaskThreadProcedure*)ice::nativeio_thread_procedure,
+                .custom_procedure_userdata = _io_thread_data.get(),
+                .debug_name = "ice.resource_tracker",
+            };
+            _io_thread = ice::create_thread(_allocator, _io_queue, io_thread_info);
         }
+    }
 
-        void refresh_providers() noexcept override
+    ResourceTrackerImplementation::~ResourceTrackerImplementation() noexcept
+    {
+        for (ice::ResourceHandle* handle : _resources)
         {
-            for (ice::ResourceProvider* provider : _providers)
+            if (handle->refcount.load(std::memory_order_relaxed) > 0)
             {
-                ice::ResourceProviderResult const result = ice::sync_wait(provider->refresh());
-                if (result == ice::ResourceProviderResult::Failure)
-                {
-                    ICE_LOG(
-                        ice::LogSeverity::Error,
-                        ice::LogTag_ResourceSystem, "Failed to update resource provider!"
-                    );
-                }
+                //IPT_MESSAGE_C("Encountered unreleased resource object during resource tracker destruction.", 0xEE99AA);
```

Comment:
Create a task to properly track resource release or at least create warnings for it in the devui.

Reply:
Created issue: https://github.com/iceshard-engine/engine/issues/158